service/history/queues/queue_base.go

```diff
@@ -0,0 +1,356 @@
+	p.checkpointTimer = time.NewTimer(backoff.JitDuration(
+		p.options.CheckpointInterval(),
+		p.options.CheckpointIntervalJitterCoefficient(),
+	))
+}
+
+func (p *queueBase) Stop() {
+	p.readerGroup.Stop()
+	p.rescheduler.Stop()
+	p.pollTimer.Stop()
+	p.checkpointTimer.Stop()
+}
+
+func (p *queueBase) Category() tasks.Category {
+	return p.category
+}
+
+func (p *queueBase) FailoverNamespace(
+	namespaceIDs map[string]struct{},
+) {
+	// TODO: reschedule all tasks for namespaces that becomes active
+	// no-op
+}
+
+func (p *queueBase) LockTaskProcessing() {
+	// no-op
+}
+
+func (p *queueBase) UnlockTaskProcessing() {
+	// no-op
+}
+
+func (p *queueBase) processPollTimer() {
+	if p.lastPollTime.Add(p.options.MaxPollInterval()).Before(p.timeSource.Now()) {
+		p.processNewRange()
+	}
+
+	p.pollTimer.Reset(backoff.JitDuration(
+		p.options.MaxPollInterval(),
+		p.options.MaxPollIntervalJitterCoefficient(),
+	))
+}
+
+func (p *queueBase) processNewRange() {
+	newMaxKey := p.shard.GetQueueExclusiveHighReadWatermark(
+		p.category,
+		p.shard.GetClusterMetadata().GetCurrentClusterName(),
+	)
+
+	if !p.nonReadableScope.CanSplitByRange(newMaxKey) {
+		return
+	}
+
+	p.lastPollTime = p.timeSource.Now()
+
+	var newReadScope Scope
+	newReadScope, p.nonReadableScope = p.nonReadableScope.SplitByRange(newMaxKey)
+	newSlice := NewSlice(
+		p.paginationFnProvider,
+		p.executableInitializer,
+		newReadScope,
+	)
+
+	reader, ok := p.readerGroup.ReaderByID(defaultReaderId)
+	if !ok {
+		p.readerGroup.NewReader(defaultReaderId, newSlice)
+	} else {
+		reader.MergeSlices(newSlice)
+	}
+}
+
+func (p *queueBase) checkpoint() {
+	var err error
+	defer func() {
+		if err == nil {
+			p.checkpointRetrier.Reset()
+			p.checkpointTimer.Reset(backoff.JitDuration(
+				p.options.CheckpointInterval(),
+				p.options.CheckpointIntervalJitterCoefficient(),
+			))
+		} else {
+			backoff := p.checkpointRetrier.NextBackOff()
+			p.checkpointTimer.Reset(backoff)
+		}
+	}()
+
+	totalSlices := 0
+	readerScopes := make(map[int32][]Scope)
+	newInclusiveLowWatermark := tasks.MaximumKey
+	for id, reader := range p.readerGroup.Readers() {
+		reader.ShrinkSlices()
+		scopes := reader.Scopes()
+
+		totalSlices += len(scopes)
+		readerScopes[id] = scopes
+		for _, scope := range scopes {
+			newInclusiveLowWatermark = tasks.MinKey(newInclusiveLowWatermark, scope.Range.InclusiveMin)
+		}
+	}
+
+	// NOTE: Must range complete task first.
+	// Otherwise, if state is updated first, later deletion fails and shard get reloaded
+	// some tasks will never be deleted.
+	if newInclusiveLowWatermark != tasks.MaximumKey && newInclusiveLowWatermark.CompareTo(p.inclusiveLowWatermark) > 0 {
+		persistenceMinTaskKey := p.inclusiveLowWatermark
+		persistenceMaxTaskKey := newInclusiveLowWatermark
+		if p.category.Type() == tasks.CategoryTypeScheduled {
+			persistenceMinTaskKey = tasks.NewKey(p.inclusiveLowWatermark.FireTime, 0)
+			persistenceMaxTaskKey = tasks.NewKey(newInclusiveLowWatermark.FireTime, 0)
```

Comment:
nit: possible to have a key.Copy()?

Reply:
Looks like it's not necessary here, just want to reset the value for taskID.